[TIMOB-26378] Android: backgroundDisabledColor is ignored

ypbnv opened the pull request:
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-26378

**Description:**
Fix the check for having a color state background for a view.

Note: There is a unit test for that, but needs some attention apparently.

**Test case:**
_app.js_
```js
var win = Ti.UI.createWindow({
    layout: 'vertical'}),
view = Ti.UI.createView({
    top: 50,
    width: 100,
    height: 100,
    backgroundColor: 'red',
    backgroundDisabledColor: 'gray'}),
button = Ti.UI.createButton({
    title: 'toggle'
});
view.addEventListener('click', function () {
    Ti.API.info('Click!');
});
button.addEventListener('click', function () {
    view.enabled = !view.enabled;
});
win.add([view, button])
win.open();
```

ypbnv committed:
Add DISABLED state in the check for color state background.

build commented:
<!--
  2 failure:  :microscope: Ther..., Tests have failed...
  0 warning: 
  1 messages
  1 markdown notices
  DangerID: danger-id-default;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :microscope: There are library changes, but no changes to the unit tests. That's OK as long as you're refactoring existing code, but will require an admin to merge this PR. Please see <a href="https://github.com/ypbnv/titanium_mobile/blob/TIMOB-26378/README.md#unit-tests">README.md#unit-tests</a> for docs on unit testing.
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  Tests have failed, see below for more information.
  </td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-10317/1/artifact/dist/mobilesdk-7.5.0.v20180911070520-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  </tbody>
</table>

### Tests: 

| Classname | Name | Time | Error |
| --- | --- | --- | --- |
| ios.Titanium.Network.HTTPClient | basic-auth success | 60.009 | file:///Users/build/Librar |

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

garymathews reviewed (approved):
CR: **PASS**

NOTE: Is this still `work in progress` ?

ypbnv committed:
Separate unit test.
Connect the getter for backgroundDisabledColor with the actual bacground drawable.

ypbnv committed:
Merge branch 'master' into TIMOB-26378

ypbnv commented:
@garymathews Yes, it was. I was trying to get the unit test working. I did it, but with a bit more code that anticipated. Let me know what you think about it. The thing that bothers me the most is the trend to limit usage of non SDK APIs in Android P ( supposing that's the way Android is taking ) which may cause the reflection to fail at some point. I have ran the tests on a Pixel with Android P and it is working fine, but that may not be the case in the future.
I have not put the code for the disabled state in TiUiHelper for now. If we decide to keep it I will move it there and add getters for DEFAULT, FOCUSED and SELECTED states as well. 
That opens the question if we can do something similar for images for different states? And if we can have the same on iOS?

CC: @jquick-axway, @hansemannn

build commented:
<!--
  0 failure: 
  0 warning: 
  1 messages
  
  DangerID: danger-id-default;
-->



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-10317/4/artifact/dist/mobilesdk-7.5.0.v20180912053015-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

garymathews commented on android/titanium/src/java/org/appcelerator/titanium/proxy/TiViewProxy.java:
```diff
@@ -1166,6 +1172,46 @@ public TiViewProxy getParent()
+				// Get the reflection methods.
+				Method getStateDrawableIndexMethod =
+					StateListDrawable.class.getMethod("getStateDrawableIndex", int[].class);
+				Method getStateDrawableMethod = StateListDrawable.class.getMethod("getStateDrawable", int.class);
```
I feel we should be a little more graceful if the methods do not exist instead of printing a stack trace

ypbnv replied:
Maybe print Warnings instead of the stack trace?

garymathews commented on android/titanium/src/java/org/appcelerator/titanium/proxy/TiViewProxy.java:
```diff
@@ -1166,6 +1172,46 @@ public TiViewProxy getParent()
+				// Get the drawable at the index.
+				Drawable drawable = (Drawable) getStateDrawableMethod.invoke(stateListDrawable, index);
+				// Parse it as a ColorDrawable.
+				ColorDrawable colorDrawable = ((ColorDrawable) ((LayerDrawable) drawable).getDrawable(0));
```
Validate we can do this cast?

garymathews commented on android/titanium/src/java/org/appcelerator/titanium/proxy/TiViewProxy.java:
```diff
@@ -1166,6 +1172,46 @@ public TiViewProxy getParent()
+				// Parse it as a ColorDrawable.
+				ColorDrawable colorDrawable = ((ColorDrawable) ((LayerDrawable) drawable).getDrawable(0));
+				// Transcript the color int to HexString.
+				String strColor = String.format("#%06X", 0xFFFFFF & colorDrawable.getColor());
```
Could this return a color with an alpha channel?

ypbnv replied:
I have missed to check that. It can, so I changed the resulting String format.

ypbnv committed:
Improve format of result to show Alpha values.
Use Warnings instead of printing stack trace.
Guard the drawable casting.

ypbnv committed:
Fix formatting.

ypbnv committed:
Fix unit test.

garymathews committed:
Merge branch 'master' into TIMOB-26378

garymathews reviewed (approved):
CR: **PASS**

ypbnv commented:
If we are OK with this, I think doing the same for the rest of the states of the color background is a good next step. What do you think ?

longton95 reviewed (approved):
**FR Passed**

`backgroundDisabledColor` is no longer ignored, The view now turns grey when disabled.

**ENV**

```
Pixel 2 XL (9.0)
Emulator (8.1, 7.1)
Appc NPM: 4.2.13
Appc CLI: 7.0.6
Ti CLI: 5.1.1
Node: 8.9.4
NPM: 6.3.0
```
